In [27]:
import pandas as pd
from scipy import stats
import numpy as np
from scipy.stats import ttest_ind, ttest_rel


### 1. One tailed t-test 

One tailed t-test - In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file files_for_lab/machine.txt. Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?

#### first solution 

In [15]:
# data for new and old machine
new_machine = [42.1, 41, 41.3, 41.8, 42.4, 42.8, 43.2, 42.3, 41.8, 42.7]

old_machine = [42.7, 43.6, 43.8, 43.3, 42.5, 43.5, 43.1, 41.7, 44, 44.1]

In [16]:
# calculate the t and p-value 

t, p = stats.ttest_ind(new_machine, old_machine, equal_var=False)

# out put the results

alpha = 0.05
print (f't = {t:.3f}')
print (f"p={p:.4f}")

if p < alpha and t < 0:
    print ("the new machine packs cartons significantly faster than the old machine.")
else:
    print("The new machine does not pack cartons significantly faster than the old machine")


t = -3.397
p=0.0032
the new machine packs cartons significantly faster than the old machine.


#### alternation solution 

In [ ]:
#Hypothesis:
# H0: μnew = μold
#H1 : μnew < μold

In [21]:
# data for new and old machine
new_machine = [42.1, 41, 41.3, 41.8, 42.4, 42.8, 43.2, 42.3, 41.8, 42.7]

old_machine = [42.7, 43.6, 43.8, 43.3, 42.5, 43.5, 43.1, 41.7, 44, 44.1]



In [22]:
# perform two_sample independent t-test

t_stat, p_value = ttest_ind(new_machine,old_machine, equal_var=False )

if p_value < 0.05:
    print ("new machine is faster")
else:
    print ('old machine is faster')

new machine is faster


### 2. Matched Pairs Test - In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). 

#### Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. 
#### Our hypothesis is that the defense and attack scores are equal. 
#### Compare the two columns to see if there is a statistically significant difference between them and comment your result.

In [33]:
#load the dataset
pokemon = pd.read_csv ('files_for_lab/pokemon.csv')
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [25]:
attack = pokemon ['Attack']
defensen = pokemon ['Defense']

In [31]:
# H0: attack = defensen 
#H1: attack != defensen 

result = ttest_rel(attack,defensen )

print('Matched Pairs Test Result:')
print('t-statistic:', result.statistic)
print('p-value:', result.pvalue)

Matched Pairs Test Result:
t-statistic: 4.325566393330478
p-value: 1.7140303479358558e-05


From the t-test result, the calculated t-statistic is 4.33 and the corresponding p-value is very small (p-value: 1.71e-05). As the p-value is smaller than the significance level of 0.05, we reject the null hypothesis and conclude that there is a statistically significant difference between attack and defense scores for the Pokemon dataset.


Therefore, we can say that there is sufficient evidence to suggest that the attack and defense scores are not equal for the Pokemon dataset

###                                              H1: attack != defensen 

## OPTIONAL PART | Inferential statistics - ANOVA

### Part 1

In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on: - Null hypothesis - Alternate hypothesis - Level of significance - Test statistic - P-value - F table

Context
Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data anova_lab_data.xlsx file in the files_for_lab folder

State the null hypothesis
State the alternate hypothesis
What is the significance level
What are the degrees of freedom of model, error terms, and total DoF
Data was collected randomly and provided to you in the table as shown

Power   					              Totals	Averages

160 w	5.43	5.71	6.22	5.59	28.96	5.792

180 w	6.24	6.71	5.56	 6.6	31.19	6.238

200 w	8.79	9.2	8.15	7.55	    41.59	8.318


					                   101.74	6.782666667

### solution 

Null hypothesis: The mean etching rates for all power levels are equal: 

Alternative hypothesis: at least one of the mean etching rates is different from the others 

Significance level: alpha = 0.05









###### Degrees of freedom:

Total degrees of freedom = (number of observations) - 1 = 11 - 1 = 10

Number of groups or levels = 3

Within groups (error) degrees of freedom = (number of groups - 1) * (number of observations per group - 1) = 2 * (4 - 1) = 6

Between groups (model) degrees of freedom = (number of groups - 1) = 2

Test statistic: The test statistic for ANOVA is the F-value, which is the ratio of the variance between groups to the variance within groups. It is calculated by dividing the mean square between groups by the mean square within groups.

P-value: The p-value is the probability of obtaining the observed test statistic (or a more extreme value) if the null hypothesis is true. If the p-value is less than the significance level, we reject the null hypothesis.

F table: The F table is used to find the critical value of F for a given significance level and degrees of freedom. We will use the F table to compare our calculated F-value to the critical value of F. If the calculated F-value is greater than the critical value, we reject the null hypothesis.

### Part 2
In this section, use the Python to conduct ANOVA.

What conclusions can you draw from the experiment and why?

In [42]:
import pandas as pd
import scipy.stats as stats

# create a DataFrame with the provided data as I cannot loading the data set by pd.read_csv function
data = {'Power': ['160 W', '180 W', '200 W', '160 W', '180 W', '200 W', '160 W', '180 W', '200 W', '160 W', '180 W', '200 W', '160 W', '180 W', '200 W'],
        'Etching Rate': [5.43, 6.24, 8.79, 5.71, 6.71, 9.2, 6.22, 5.98, 7.9, 6.01, 5.66, 8.15, 5.59, 6.6, 7.55]}
df = pd.DataFrame(data)

# conduct ANOVA
result = stats.f_oneway(df[df['Power']=='160 W']['Etching Rate'],
                        df[df['Power']=='180 W']['Etching Rate'],
                        df[df['Power']=='200 W']['Etching Rate'])

# print ANOVA results
print('One-way ANOVA')
print('F-value:', result.statistic)
print('p-value:', result.pvalue)



One-way ANOVA
F-value: 36.87895470100505
p-value: 7.506584272358903e-06


The F-value is 36.879 

the p-value is 7.507 x 10^-6, which is less than the commonly used significance level of 0.05.

This means that we can reject the null hypothesis that the means of etching rate at different power levels are equal, and conclude that at least one power level has a significantly different etching rate mean than the others.